daemon/cmd/state.go

```diff
@@ -304,59 +301,37 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 		}
 	}
 
-	if option.Config.EnableIPSec {
-		// To support v1.18 VinE upgrades, we need to restore the host
-		// endpoint before any other endpoint, to ensure a drop-less upgrade.
-		// This is because in v1.18 'bpf_lxc' programs stop issuing IPsec hooks
-		// which trigger encryption.
-		//
-		// Instead, 'bpf_host' is responsible for performing IPsec hooks.
-		// Therefore, we want 'bpf_host' to regenerate BEFORE 'bpf_lxc' so the
-		// IPsec hooks are always present while 'bpf_lxc' programs regen,
-		// ensuring no IPsec leaks occur.
-		//
-		// This can be removed in v1.19.
-		for _, ep := range state.restored {
-			if ep.IsHost() {
-				d.logger.Info("Successfully restored endpoint. Scheduling regeneration", logfields.EndpointID, ep.ID)
-				if err := ep.RegenerateAfterRestore(endpointsRegenerator, d.endpointMetadata.FetchK8sMetadataForEndpoint); err != nil {
-					d.logger.Debug(
-						"error regenerating during restore",
-						logfields.Error, err,
-						logfields.EndpointID, ep.ID,
-					)
-					epRegenerated <- false
-				} else {
-					epRegenerated <- true
-				}
-				break
-			}
-		}
-	}
-
+	endpointsToRegenerate := make([]*endpoint.Endpoint, 0, len(state.restored))
 	for _, ep := range state.restored {
 		if ep.IsHost() && option.Config.EnableIPSec {
-			// The host endpoint was handled above.
-			continue
-		}
-		d.logger.Info(
-			"Successfully restored endpoint. Scheduling regeneration",
-			logfields.EndpointID, ep.ID,
-		)
-		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
+			// To support v1.18 VinE upgrades, we need to restore the host
+			// endpoint before any other endpoint, to ensure a drop-less upgrade.
+			// This is because in v1.18 'bpf_lxc' programs stop issuing IPsec hooks
+			// which trigger encryption.
+			//
+			// Instead, 'bpf_host' is responsible for performing IPsec hooks.
+			// Therefore, we want 'bpf_host' to regenerate BEFORE 'bpf_lxc' so the
+			// IPsec hooks are always present while 'bpf_lxc' programs regen,
+			// ensuring no IPsec leaks occur.
+			//
+			// This can be removed in v1.19.
+			d.logger.Info("Successfully restored Host endpoint. Scheduling regeneration", logfields.EndpointID, ep.ID)
 			if err := ep.RegenerateAfterRestore(endpointsRegenerator, d.endpointMetadata.FetchK8sMetadataForEndpoint); err != nil {
 				d.logger.Debug(
-					"error regenerating during restore",
+					"Error regenerating Host endpoint during restore",
 					logfields.Error, err,
 					logfields.EndpointID, ep.ID,
 				)
-				epRegenerated <- false
-				return
 			}
-			epRegenerated <- true
-		}(ep, epRegenerated)
+			continue
+		}
+
+		endpointsToRegenerate = append(endpointsToRegenerate, ep)
 	}
 
+	// Trigger regeneration for relevant restored endpoints in a separate goroutine.
+	go d.handleRestoredEndpointsRegeneration(endpointsToRegenerate, endpointsRegenerator)
```

Comment:
There's a subtle reordering of logic here in the last patch that I'm not sure I understand the repercussions for: Previously, the current goroutine would wait until all endpoints were removed before beginning the goroutine that waits for full policy computation to complete. This meant that `d.endpointInitialPolicyCompute` would only be closed once all restoration is complete, including both policy calculation for live endpoints as well as the removal of endpoints that were selected for cleanup/removal.

Is this change intentional? Is it safe? If in doubt, I would avoid changing the order (or do so as a separate change, maybe even not part of this bugfix).

Reply:
Ah ok, yeah you are right. We should keep the old order of dependencies between endpoint cleanup and restore complete channel close. Should just need reordering, will make the fix.

Reply:
Thanks, LGTM.